File: packages/node/iii-example/src/hooks.ts
import { type ApiRequest, type ApiResponse, type Context, getContext } from 'iii-sdk'
import { iii } from './iii'

export const useApi = (
  config: {
    api_path: string
    http_method: string
    description?: string
    metadata?: Record<string, any>
  },
  handler: (req: ApiRequest<any>, context: Context) => Promise<ApiResponse>,
) => {
  const function_id = `api.${config.http_method.toLowerCase()}.${config.api_path}`

  iii.registerFunction({ id: function_id, metadata: config.metadata }, req =>
    handler(req, getContext()),
  )
  iii.registerTrigger({
    trigger_type: 'api',
    function_id,
    config: {
      api_path: config.api_path,
      http_method: config.http_method,
      description: config.description,
      metadata: config.metadata,
    },
  })
}

File: packages/python/iii-example/src/hooks.py
from typing import Any, Awaitable, Callable

from iii import ApiRequest, ApiResponse, FunctionInfo, get_context

from .iii import iii


def use_api(
    config: dict[str, Any],
    handler: Callable[[ApiRequest[Any], Any], Awaitable[ApiResponse[Any]]],
) -> None:
    api_path = config["api_path"]
    http_method = config["http_method"]
    function_id = f"api.{http_method.lower()}.{api_path}"

    async def wrapped(data: Any) -> dict[str, Any]:
        req = ApiRequest(**data) if isinstance(data, dict) else data
        ctx = get_context()
        result = await handler(req, ctx)
        return result.model_dump(by_alias=True)

    iii.register_function(function_id, wrapped)
    iii.register_trigger(
        trigger_type="api",
        function_id=function_id,
        config={
            "api_path": api_path,
            "http_method": http_method,
            "description": config.get("description"),
            "metadata": config.get("metadata"),
        },
    )


def use_functions_available(callback: Callable[[list[FunctionInfo]], None]) -> Callable[[], None]:
    return iii.on_functions_available(callback)
